File: shared/src/main/java/ru/privatenull/pnauth/extension/AuthPolicyDecision.java
package ru.privatenull.pnauth.extension;
import java.time.Duration;
public record AuthPolicyDecision(Type type, String provider, String message, Duration lifetime) {
    public enum Type { ALLOW, DENY, REQUIRE_VERIFICATION }
    public static AuthPolicyDecision allow() { return new AuthPolicyDecision(Type.ALLOW, "", "", Duration.ZERO); }
    public static AuthPolicyDecision deny(String message) { return new AuthPolicyDecision(Type.DENY, "", message, Duration.ZERO); }
    public static AuthPolicyDecision requireVerification(String provider, String message, Duration lifetime) {
        if (provider == null || provider.isBlank()) throw new IllegalArgumentException("provider is required");
        return new AuthPolicyDecision(Type.REQUIRE_VERIFICATION, provider, message,
                lifetime == null || lifetime.isNegative() || lifetime.isZero() ? Duration.ofMinutes(5) : lifetime);
    }
}

File: shared/src/main/kotlin/ru/privatenull/pnauth/extension/AuthPolicyDecision.kt
package ru.privatenull.pnauth.extension

import java.time.Duration

@JvmRecord
data class AuthPolicyDecision(
    val type: Type,
    val provider: String,
    val message: String,
    val lifetime: Duration
) {
    enum class Type { ALLOW, DENY, REQUIRE_VERIFICATION }

    companion object {
        @JvmStatic
        fun allow(): AuthPolicyDecision = AuthPolicyDecision(Type.ALLOW, "", "", Duration.ZERO)

        @JvmStatic
        fun deny(message: String): AuthPolicyDecision = AuthPolicyDecision(Type.DENY, "", message, Duration.ZERO)

        @JvmStatic
        fun requireVerification(provider: String?, message: String, lifetime: Duration?): AuthPolicyDecision {
            require(!provider.isNullOrBlank()) { "provider is required" }
            val effectiveLifetime = if (lifetime == null || lifetime.isNegative || lifetime.isZero) {
                Duration.ofMinutes(5)
            } else {
                lifetime
            }
            return AuthPolicyDecision(Type.REQUIRE_VERIFICATION, provider, message, effectiveLifetime)
        }
    }
}
